File: services/clinical-service/src/domain/value-objects/Diagnosis.ts
export class Diagnosis {
    private readonly code: string;
    private readonly description: string;

    private constructor(code: string, description: string) {
        this.code = code;
        this.description = description;
    }

    static create(code: string, description: string): Diagnosis {
        if (!code || code.trim().length === 0) {
            throw new Error('Diagnosis code cannot be empty');
        }

        // ICD-10 format: letter followed by 2+ digits, optional dot and more digits
        const icd10Pattern = /^[A-Z]\d{2}(\.\d{1,4})?$/;
        if (!icd10Pattern.test(code)) {
            throw new Error(
                'Diagnosis code must be a valid ICD-10 format, e.g. J06.9'
            );
        }

        if (!description || description.trim().length === 0) {
            throw new Error('Diagnosis description cannot be empty');
        }

        return new Diagnosis(code, description);
    }

    getCode(): string {
        return this.code;
    }

    getDescription(): string {
        return this.description;
    }

    equals(other: Diagnosis): boolean {
        return this.code === other.code;
    }

    toString(): string {
        return `${this.code} - ${this.description}`;
    }
}

File: services/clinical-service/src/main/java/com/medtrust/clinical/domain/model/valueobject/Diagnosis.java
package com.medtrust.clinical.domain.model.valueobject;

import java.util.Objects;
import java.util.regex.Pattern;

public final class Diagnosis {

    private static final Pattern ICD10_PATTERN = Pattern.compile("^[A-Z]\\d{2}(\\.\\d{1,4})?$");

    private final String code;
    private final String description;

    private Diagnosis(String code, String description) {
        this.code = code;
        this.description = description;
    }

    public static Diagnosis create(String code, String description) {
        if (code == null || code.isBlank()) {
            throw new IllegalArgumentException("Diagnosis code cannot be empty");
        }
        if (!ICD10_PATTERN.matcher(code).matches()) {
            throw new IllegalArgumentException(
                    "Diagnosis code must be a valid ICD-10 format, e.g. J06.9");
        }
        if (description == null || description.isBlank()) {
            throw new IllegalArgumentException("Diagnosis description cannot be empty");
        }
        return new Diagnosis(code, description);
    }

    public String getCode() {
        return code;
    }

    public String getDescription() {
        return description;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (o == null || getClass() != o.getClass())
            return false;
        Diagnosis diagnosis = (Diagnosis) o;
        return Objects.equals(code, diagnosis.code);
    }

    @Override
    public int hashCode() {
        return Objects.hash(code);
    }

    @Override
    public String toString() {
        return code + " - " + description;
    }
}
